src/auth0-sso-login.js

```diff
@@ -6,13 +6,55 @@ export class windowInteraction {
   static updateWindow(url) {
     window.location = url;
   }
+
+  static setTimeout(func, delay) {
+    return window.setTimeout(func, delay);
+  }
+
+  static clearTimeout(timeoutId) {
+    window.clearTimeout(timeoutId);
+  }
 }
 
+// static localStorage interaction, mostly wrapped for enabling easier mocking through unit tests
+export class localStorageInteraction {
+  static getItem(key) {
+    return localStorage.getItem(key);
+  }
+
+  static setItem(key, value) {
+    return localStorage.setItem(key, value);
+  }
+
+  static removeItem(key) {
+    return localStorage.removeItem(key);
+  }
+}
+
+export const tokenExpiresAtKey = 'cimpress.auth0.tokenExpiresAt';
+
+const scheduleRenewal = (tokenRefreshHandle, renewFunction) => {
+  const tokenExpiresAtJson = localStorageInteraction.getItem(tokenExpiresAtKey);
+  if (!tokenExpiresAtJson) {
+    return null;
+  }
+  const tokenExpiresAt = JSON.parse(tokenExpiresAtJson);
+
+  if (tokenRefreshHandle) {
+    windowInteraction.clearTimeout(tokenRefreshHandle);
+  }
+
+  // refresh token in the second third of its lifetime
+  const refreshDelay = Math.floor((tokenExpiresAt - Date.now()) / 3) * 2;
+  return windowInteraction.setTimeout(renewFunction, refreshDelay);
```

Comment:
With `window.setTimeout` If the refresh fails `renewFunction` will never get called a second time.  Rather than rescheduling the refresh as part of the `renumFunction` I suggest using `setInterval` here, and removing the line
`this.tokenRefreshHandle = scheduleRenewal(this.tokenRefreshHandle, () => this.ensureLoggedIn());` from the `tokenRefresh` call.

Reply:
If it fails, the resulting state should be auth0 lock open waiting for user interaction. There is no reason to have the refresh interval still running.

If we find out that `setInterval` will be more reliable, we should make sure that it will not open multiple auth0 locks as it fires while user is entering his credentials.